src/pools/D3MAaveDaiPool.sol

```diff
@@ -0,0 +1,176 @@
+        require(variableDebt_ != address(0), "D3MAaveDaiPool/invalid-variableDebt");
+        require(interestStrategy_ != address(0), "D3MAaveDaiPool/invalid-interestStrategy");
+
+        stableDebt = ShareTokenLike(stableDebt_);
+        variableDebt = ShareTokenLike(variableDebt_);
+        interestStrategy = InterestRateStrategyLike(interestStrategy_);
+        rewardsClaimer = RewardsClaimerLike(_rewardsClaimer);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        ShareTokenLike(adai_).approve(address(pool_), type(uint256).max);
+        TokenLike(dai_).approve(address(pool_), type(uint256).max);
+
+    }
+
+    // --- Math ---
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "D3MAaveDaiPool/overflow");
+    }
+    function _sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, "D3MAaveDaiPool/underflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "D3MAaveDaiPool/overflow");
+    }
+    uint256 constant RAY  = 10 ** 27;
+    function _rmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, y) / RAY;
+    }
+    function _rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, RAY) / y;
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Admin ---
+    function file(bytes32 what, address data) public override auth {
+        require(live == 1, "D3MAaveDaiPoolPool/no-file-not-live");
+
+        if (what == "king") king = data;
+        else super.file(what, data);
+    }
+
+    function validTarget() external view override returns (bool) {
+        (,,,,,,,,,, address strategy,) = LendingPoolLike(pool).getReserveData(address(asset));
+        return strategy == address(interestStrategy);
+    }
+
+    function calcSupplies(uint256 availableAssets) external view override returns(uint256, uint256) {
+        return D3MPlanLike(plan).calcSupplies(availableAssets);
+    }
+
+    // Deposits Dai to Aave in exchange for adai which gets sent to the msg.sender
+    // Aave: https://docs.aave.com/developers/v/2.0/the-core-protocol/lendingpool#deposit
+    function deposit(uint256 amt) external override auth {
+        LendingPoolLike(pool).deposit(address(asset), amt, address(this), 0);
+
+    }
+
+    // Withdraws Dai from Aave in exchange for adai
+    // Aave: https://docs.aave.com/developers/v/2.0/the-core-protocol/lendingpool#withdraw
+    function withdraw(uint256 amt) external override auth {
+        LendingPoolLike(pool).withdraw(address(asset), amt, address(this));
+    }
+
+    // --- Collect any rewards ---
+    function collect(address[] memory assets, uint256 amount) external override auth returns (uint256 amt) {
```

Comment:
On compound you can't specify an exact amount to collect/claim:
https://github.com/compound-finance/compound-protocol/blob/master/contracts/ComptrollerG7.sol#L1215

Suggest to change general functionality to claim all rewards we are entitled to. ANothe option would be to ignore the `amount` parameter in compound (or  verify it was given as 0), but I think those are less preferable.

In any case `amount` and `amt` are confusing. We should probably not use `amount` at all as `amt` is used a lot in this contract.

Reply:
Mmm also pretty annoying that in AAVE you need to specify assets in their contract (i.e underlying tokens) and in Compound you specify ctokens (i.e the share tokens).
So it seems we should either change `assets` to something more vague like `tokens`, or change to `shares`  here, and in our AAVE implementation loop and convert each share to the underlying asset before passing it further.
Not sure which one is better, maybe the `tokens` solution since it's just simpler.

BTW Compound also has a more generic function that does not need an array, but seems to be really not efficient:
https://github.com/compound-finance/compound-protocol/blob/master/contracts/ComptrollerG7.sol#L1206

Reply:
As I said above rewards are optional, platform-specific and usually pretty bespoke so I think we should move them entirely into the pool and have no concept of rewards, collect, king, etc in the Hub.

Reply:
Yeah, works. I was hesitant to suggest that since up untill now the pool was not a user facing contract so far but I thinj it's fune.